python/ray/data/datasource/path_util.py

```diff
@@ -42,13 +133,87 @@ def _has_file_extension(path: str, extensions: Optional[List[str]]) -> bool:
     return any(path.lower().endswith(ext) for ext in extensions)
 
 
+def _resolve_single_path_with_fallback(
+    path: str,
+    filesystem: Optional["pyarrow.fs.FileSystem"] = None,
+) -> Tuple["pyarrow.fs.FileSystem", str]:
+    """Resolve a single path with filesystem, with fallback to re-resolution on error.
+
+    This is a helper for lazy filesystem resolution. If a filesystem is provided,
+    it first attempts to validate the path against that filesystem. If that fails
+    (e.g., the path uses a different filesystem type), it re-resolves without
+    the cached filesystem.
+
+    Args:
+        path: A single file/directory path.
+        filesystem: Optional cached filesystem from previous resolution.
+
+    Returns:
+        Tuple of (resolved_filesystem, resolved_path).
+
+    Raises:
+        ValueError: If path resolution fails.
+        ImportError: If required dependencies are missing.
+    """
+    import pyarrow as pa
+    from pyarrow.fs import _resolve_filesystem_and_path
+
+    path = _resolve_custom_scheme(path)
+
+    # Validate/wrap filesystem if needed
+    try:
+        filesystem = _validate_and_wrap_filesystem(filesystem)
+    except TypeError as e:
+        raise ValueError(f"Invalid filesystem provided: {e}") from e
+
+    # Check HTTP scheme FIRST - PyArrow doesn't support HTTP/HTTPS natively
+    parsed = urlparse(path, allow_fragments=False)
+    if parsed.scheme in ["http", "https"]:
+        try:
+            resolved_filesystem = _get_fsspec_http_filesystem()
+            resolved_path = path
+            return resolved_filesystem, resolved_path
+        except ImportError as import_error:
+            raise ImportError(
+                f"Cannot resolve HTTP path '{path}': {import_error}"
+            ) from import_error
```

Comment:
Previously, I think we'd only create a HTTP filesystem if `_resolve_filesystem_and_path` failed, whereas in this implementation I think we unconditionally create a HTTP filesystem. 

Are there scenarios where the paths are HTTP and `_resolve_filesystem_and_path` _doesn't_ error? (e.g., if you pass in a custom filesystem that works with HTTP?) If so, I think this could be a breaking change